Tools/LyTestTools/ly_test_tools/benchmark/data_aggregator.py

```diff
@@ -0,0 +1,163 @@
+
+from ly_test_tools.mars.filebeat_client import FilebeatClient
+
+class BenchmarkPathException(Exception):
+    """Custom Exception class for invalid benchmark file paths."""
+    pass
+
+class BenchmarkDataAggregator(object):
+    def __init__(self, workspace, logger, test_suite):
+        self.build_dir = workspace.paths.build_directory()
+        self.results_dir = Path(workspace.paths.project(), 'user/Scripts/PerformanceBenchmarks')
+        self.test_suite = test_suite
+        self.filebeat_client = FilebeatClient(logger)
+
+    def _update_pass(self, pass_stats, entry):
+        '''
+        Modifies pass_stats dict keyed by pass name with the time recorded in a pass timestamp entry.
+
+        :param pass_stats: dict aggregating statistics from each pass (key: pass name, value: dict with stats)
+        :param entry: dict representing the timestamp entry of a pass
+        :return: Time (in nanoseconds) recorded by this pass
+        '''
+        name = entry['passName']
+        time_ns = entry['timestampResultInNanoseconds']
+        pass_entry = pass_stats.get(name, { 'totalTime': 0, 'maxTime': 0 })
+
+        pass_entry['maxTime'] = max(time_ns, pass_entry['maxTime'])
+        pass_entry['totalTime'] += time_ns
+        pass_stats[name] = pass_entry
+        return time_ns
+
+
+    def _process_benchmark(self, benchmark_dir, benchmark_metadata):
+        '''
+        Aggregates data from results from a single benchmark contained in a subdirectory of self.results_dir.
+
+        :param benchmark_dir: Path of directory containing the benchmark results
+        :param benchmark_metadata: Dict with benchmark metadata mutated with additional info from metadata file
+        :return: Tuple with two indexes:
+            [0]: Dict aggregating statistics from frame times (key: stat name)
+            [1]: Dict aggregating statistics from pass times (key: pass name, value: dict with stats)
+        '''
+        # Parse benchmark metadata
+        metadata_file = benchmark_dir / 'benchmark_metadata.json'
+        if metadata_file.exists():
+            data = json.loads(metadata_file.read_text())
+            benchmark_metadata.update(data['ClassData'])
+        else:
+            raise BenchmarkPathException(f'Metadata file could not be found at {metadata_file}')
+
+        # data structures aggregating statistics from timestamp logs
+        frame_stats = { 'count': 0, 'totalTime': 0, 'maxTime': 0, 'minTime': float('inf') }
+        pass_stats = {}  # key: pass name, value: dict with totalTime and maxTime keys
+
+        # this allows us to add additional data if necessary, e.g. frame_test_timestamps.json
+        is_timestamp_file = lambda file: file.name.startswith('frame') and file.name.endswith('_timestamps.json')
+
+        # parse benchmark files
+        for file in benchmark_dir.iterdir():
+            if file.is_dir() or not is_timestamp_file(file):
+                continue
+
+            data = json.loads(file.read_text())
```

Comment:
Can this use a context manager instead to read the json file?

Reply:
It already does. This is pathlib's implementation of `read_text` (see https://github.com/python/cpython/blob/3.5/Lib/pathlib.py#L1160):
```py
def read_text(self, encoding=None, errors=None):
    """
    Open the file in text mode, read it, and close the file.
    """
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
        return f.read()
 ```